File: services/about/cpp/src/AboutService.cc
/******************************************************************************
 * Copyright (c) 2013-2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <algorithm>
#include <qcc/Debug.h>
#include <alljoyn/about/AboutService.h>
#include <alljoyn/BusAttachment.h>
#include <alljoyn/AllJoynStd.h>

#define QCC_MODULE "ALLJOYN_ABOUT_SERVICE"
#define ABOUT_SERVICE_VERSION 1

using namespace ajn;
using namespace services;

static const char* ABOUT_INTERFACE_NAME = "org.alljoyn.About";

AboutService::AboutService(ajn::BusAttachment& bus, PropertyStore& store) :
    BusObject("/About"), m_BusAttachment(&bus), m_PropertyStore(&store),
    m_AnnounceSignalMember(NULL), m_AnnouncePort(0) {

    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    std::vector<qcc::String> v;
    v.push_back(ABOUT_INTERFACE_NAME);
    m_announceObjectsLock.Lock(MUTEX_CONTEXT);
    m_AnnounceObjectsMap.insert(std::pair<qcc::String, std::vector<qcc::String> >("/About", v));
    m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
}

QStatus AboutService::Register(int port) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_OK;

    m_AnnouncePort = port;
    const InterfaceDescription* p_InterfaceDescription = NULL;
    if (m_BusAttachment) {
        p_InterfaceDescription = m_BusAttachment->GetInterface(org::alljoyn::About::InterfaceName);
    }

    if (!p_InterfaceDescription) {
        return ER_BUS_CANNOT_ADD_INTERFACE;
    }
    status = AddInterface(*p_InterfaceDescription);

    if (status == ER_OK) {
        status = AddMethodHandler(p_InterfaceDescription->GetMember("GetAboutData"),
                                  static_cast<MessageReceiver::MethodHandler>(&AboutService::GetAboutData));
        if (status != ER_OK) {
            return status;
        }
        status = AddMethodHandler(p_InterfaceDescription->GetMember("GetObjectDescription"),
                                  static_cast<MessageReceiver::MethodHandler>(&AboutService::GetObjectDescription));
        if (status != ER_OK) {
            return status;
        }

        m_AnnounceSignalMember = p_InterfaceDescription->GetMember("Announce");
        assert(m_AnnounceSignalMember);
    }

    return (status == ER_BUS_IFACE_ALREADY_EXISTS) ? ER_OK : status;
}

void AboutService::Unregister() {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
}

QStatus AboutService::AddObjectDescription(qcc::String const& path, std::vector<qcc::String> const& interfaceNames) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_OK;
    m_announceObjectsLock.Lock(MUTEX_CONTEXT);
    std::map<qcc::String, std::vector<qcc::String> >::iterator it = m_AnnounceObjectsMap.find(path);
    if (it != m_AnnounceObjectsMap.end()) {
        it->second.insert(it->second.end(), interfaceNames.begin(), interfaceNames.end());
    } else {
        m_AnnounceObjectsMap.insert(std::pair<qcc::String, std::vector<qcc::String> >(path, interfaceNames));
    }
    m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
    return status;
}

QStatus AboutService::RemoveObjectDescription(qcc::String const& path, std::vector<qcc::String> const& interfaceNames) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_OK;
    m_announceObjectsLock.Lock(MUTEX_CONTEXT);
    std::map<qcc::String, std::vector<qcc::String> >::iterator it = m_AnnounceObjectsMap.find(path);
    if (it != m_AnnounceObjectsMap.end()) {
        for (std::vector<qcc::String>::const_iterator itv = interfaceNames.begin(); itv != interfaceNames.end(); ++itv) {
            std::vector<qcc::String>::iterator findIterator = std::find(it->second.begin(), it->second.end(), itv->c_str());
            if (findIterator != it->second.end()) {
                it->second.erase(findIterator);
            }
        }
        if (it->second.empty()) {
            m_AnnounceObjectsMap.erase(it);
        }
    }
    m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
    return status;
}

QStatus AboutService::Announce() {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_OK;
    if (m_AnnounceSignalMember == NULL) {
        return ER_FAIL;
    }
    MsgArg announceArgs[4];
    status = announceArgs[0].Set("q", ABOUT_SERVICE_VERSION);
    if (status != ER_OK) {
        return status;
    }
    status = announceArgs[1].Set("q", m_AnnouncePort);
    if (status != ER_OK) {
        return status;
    }

    m_announceObjectsLock.Lock(MUTEX_CONTEXT);
    std::vector<MsgArg> announceObjectsArg(m_AnnounceObjectsMap.size());
    int objIndex = 0;
    for (std::map<qcc::String, std::vector<qcc::String> >::const_iterator it = m_AnnounceObjectsMap.begin();
         it != m_AnnounceObjectsMap.end(); ++it) {

        qcc::String objectPath = it->first;
        std::vector<const char*> interfacesVector(it->second.size());
        std::vector<qcc::String>::const_iterator interfaceIt;
        int interfaceIndex = 0;

        for (interfaceIt = it->second.begin(); interfaceIt != it->second.end(); ++interfaceIt) {
            interfacesVector[interfaceIndex++] = interfaceIt->c_str();
        }

        status = announceObjectsArg[objIndex].Set("(oas)", objectPath.c_str(), interfacesVector.size(), (interfacesVector.empty()) ? NULL : &interfacesVector.front());
        if (status != ER_OK) {
            m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
            return status;
        }
        objIndex++;
    }
    m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
    status = announceArgs[2].Set("a(oas)", objIndex, (announceObjectsArg.empty()) ? NULL : &announceObjectsArg.front());
    if (status != ER_OK) {
        return status;
    }
    status = m_PropertyStore->ReadAll(NULL, PropertyStore::ANNOUNCE, announceArgs[3]);
    if (status != ER_OK) {
        return status;
    }
    Message msg(*m_BusAttachment);
    uint8_t flags = ALLJOYN_FLAG_SESSIONLESS;
#if !defined(NDEBUG)
    for (int i = 0; i < 4; i++) {
        QCC_DbgPrintf(("announceArgs[%d]=%s", i, announceArgs[i].ToString().c_str()));
    }
#endif
    status = Signal(NULL, 0, *m_AnnounceSignalMember, announceArgs, 4, (unsigned char) 0, flags);

    QCC_DbgPrintf(("Sent AnnounceSignal from %s  =%d", m_BusAttachment->GetUniqueName().c_str(), status));
    return status;

}

void AboutService::GetAboutData(const ajn::InterfaceDescription::Member* member, ajn::Message& msg) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_OK;
    const ajn::MsgArg* args = 0;
    size_t numArgs = 0;
    msg->GetArgs(numArgs, args);
    if (numArgs == 1) {
        ajn::MsgArg retargs[1];
        status = m_PropertyStore->ReadAll(args[0].v_string.str, PropertyStore::READ, retargs[0]);
        QCC_DbgPrintf(("m_pPropertyStore->ReadAll(%s,PropertyStore::READ)  =%s", args[0].v_string.str, QCC_StatusText(status)));
        if (status != ER_OK) {
            if (status == ER_LANGUAGE_NOT_SUPPORTED) {
                MethodReply(msg, "org.alljoyn.Error.LanguageNotSupported", "The language specified is not supported");
                return;
            }
            MethodReply(msg, status);
            return;
        } else {
            MethodReply(msg, retargs, 1);
        }
    } else {
        MethodReply(msg, ER_INVALID_DATA);
    }
}

void AboutService::GetObjectDescription(const ajn::InterfaceDescription::Member* member, ajn::Message& msg) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    const ajn::MsgArg* args = 0;
    size_t numArgs = 0;
    msg->GetArgs(numArgs, args);
    if (numArgs == 0) {
        ajn::MsgArg retargs[1];
        m_announceObjectsLock.Lock(MUTEX_CONTEXT);
        std::vector<MsgArg> objectArg(m_AnnounceObjectsMap.size());
        int objIndex = 0;
        for (std::map<qcc::String, std::vector<qcc::String> >::const_iterator it = m_AnnounceObjectsMap.begin();
             it != m_AnnounceObjectsMap.end(); ++it) {
            const qcc::String& key = it->first;
            std::vector<const char*> interfacesVec(it->second.size());
            std::vector<qcc::String>::const_iterator interfaceIt;
            int interfaceIndex = 0;

            for (interfaceIt = it->second.begin(); interfaceIt != it->second.end(); ++interfaceIt) {
                interfacesVec[interfaceIndex++] = interfaceIt->c_str();
            }

            objectArg[objIndex].Set("(oas)", key.c_str(), interfacesVec.size(), (interfacesVec.empty()) ? NULL : &interfacesVec.front());
            objIndex++;
        }
        m_announceObjectsLock.Unlock(MUTEX_CONTEXT);
        retargs[0].Set("a(oas)", objectArg.size(), (objectArg.empty()) ? NULL : &objectArg.front());
        MethodReply(msg, retargs, 1);
    } else {
        MethodReply(msg, ER_INVALID_DATA);
    }
}

QStatus AboutService::Get(const char*ifcName, const char*propName, MsgArg& val) {
    QCC_DbgTrace(("AboutService::%s", __FUNCTION__));
    QStatus status = ER_BUS_NO_SUCH_PROPERTY;
    if (0 == strcmp(ifcName, ABOUT_INTERFACE_NAME)) {
        if (0 == strcmp("Version", propName)) {
            status = val.Set("q", ABOUT_SERVICE_VERSION);
        }
    }
    return status;
}

File: services/about/java/src/org/alljoyn/about/AboutService.java
/******************************************************************************
 * Copyright (c) 2013-2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

package org.alljoyn.about;

import java.util.List;

import org.alljoyn.about.client.AboutClient;
import org.alljoyn.about.icon.AboutIconClient;
import org.alljoyn.about.transport.AboutTransport;
import org.alljoyn.bus.BusAttachment;
import org.alljoyn.bus.BusException;
import org.alljoyn.services.common.AnnouncementHandler;
import org.alljoyn.services.common.BusObjectDescription;
import org.alljoyn.services.common.PropertyStore;
import org.alljoyn.services.common.ServiceAvailabilityListener;
import org.alljoyn.services.common.ServiceCommon;

/**
 * An interface for both About client (consumer) and server (producer).
 * An application may want to implement both, but still use one bus, so for
 * convenience both functionalities are encapsulated here.
 */
public interface AboutService extends ServiceCommon
{
    /**
     * The About protocol version.
     */
    public static final int PROTOCOL_VERSION = 1;

    /**
     * The About AllJoyn BusInterface name
     */
    public static final String ANNOUNCE_IFNAME   = AboutTransport.INTERFACE_NAME;

    /**
     * The Announcement signal name
     */
    public static final String SIGNAL_NAME       = "Announce";

    /**
     * Starts the {@link AboutService} in client mode.
     * @param bus the AllJoyn bus attachment
     * @throws Exception Is thrown if a failure has occurred while starting the client
     */
    public void startAboutClient(BusAttachment bus) throws Exception;

    /**
     * Creates an About client for a peer to receive Announcement signals and
     * to call remote methods.
     * @param peerName The bus unique name of the remote About server
     * @param serviceAvailabilityListener listener for connection loss
     * @param port the peer's bound port of the About server
     * @return AboutClient to create a session with the peer
     * @throws Exception indicating failure to create an AboutClient
     */
    public AboutClient createAboutClient(String peerName, ServiceAvailabilityListener serviceAvailabilityListener, short port) throws Exception;

    /**
     * Create an Icon client for a peer.
     * @param peerName The bus unique name of the remote About server
     * @param serviceAvailabilityListener listener for connection loss
     * @param port the peer's bound port of the About server
     * @return AboutIconClient to create a session with the peer
     * @throws BusException indicating failure to creat an AboutIconClient
     */
    public AboutIconClient createAboutIconClient(String peerName, ServiceAvailabilityListener serviceAvailabilityListener, short port) throws BusException;

    /**
     * Stops client mode. Disconnect all sessions.
     * @throws Exception indicating issue when stopping the AboutClient
     */
    public void stopAboutClient() throws Exception;

    /**
     * <p>Registers a handler to receive announcements.</p>
     *
     * <p>The handler is only called if all the interfaces are implemented.
     * For example, if the handler should be called if both "com.example.Audio"
     * <i>and</i> "com.example.Video" are implemented then call</p>
     * addAnnouncementHandler once:
     * <pre>
     * {@code
     * addAnnouncementHandler(handler, new String[] {"com.example.Audio", "com.example.Video"});
     * }</pre>
     *
     * <p>If the handler should be called if "com.example.Audio" <i>or</i>
     * "com.example.Video" is implemented then call</p>
     * addAnnouncementHandler multiple times:
     * <pre>
     * {@code
     * addAnnouncementHandler(handler, new String[] {"com.example.Audio"});
     * addAnnouncementHandler(handler, new String[] {"com.example.Video"});
     * }</pre>
     *
     * <p>The interface name may be a prefix followed by a <code>*</code>.  Using
     * this, the example above could be written as:</p>
     * <pre>
     * {@code
     * addAnnouncementHandler(handler, new String[] {"com.example.*"});
     * }</pre>
     * <p>The handler will receive any announcement that implements an interface
     * beginning with the "com.example." name.</p>
     *
     * <p>If RegisterAnnounceHandler is called with lists of interfaces that overlap
     * then multiple Announcements will occur.</p>
     * For example given the following:
     * <pre>
     * {@code
     * addAnnouncementHandler(handler, new String[] {"com.example.Audio"});
     * addAnnouncementHandler(handler, new String[] {"com.example.Audio", "com.example.Video"});
     * }</pre>
     *
     * <p>If a BusAttachment is found that implements both `com.example.Audio`  and
     * interface `com.example.Video` the announce handler will be called twice.
     * Once for the first added AnnouncementHandler for `com.example.Audio` and
     * again for the second added AnnouncementHandler that is looking for both
     * audio and video interfaces.</p>
     *
     * <p>If the same handler is used for for multiple interfaces then it is the
     * handlers responsibility to parse through the reported interfaces to
     * figure out what should be done in response to the Announce signal.</p>
     *
     * <p><em>Note:</em> specifying null for the interfaces parameter could have
     * significant impact on network performance and should be avoided unless
     * all announcements are needed.</p>
     *
     * @param handler the AnnouncementHandler that will be receiving the
     *                announcement.
     * @param interfaces an array of interfaces the remote service must
     *                   implement to receive the announcement.  If this
     *                   parameter is <code>null</code> the handler will receive
     *                   all announcements.
     *
     * @see AboutTransport#Announce(short, short, BusObjectDescription[], java.util.Map)
     */
    public void addAnnouncementHandler(AnnouncementHandler handler, String[] interfaces);

    /**
     * Unregisters an announcement handler.
     *
     * When calling this the array of interfaces must match the array that was
     * used when calling <code>addAnnouncementHander</code>.
     *
     * @param handler the AnnouncementHandler to remove. The array of interfaces
     *                must be the same as the array used when calling
     *                addAnnouncementHandler.
     * @param interfaces an array of interfaces that were used when calling
     *                   addAnnouncementHandler
     */
    public void removeAnnouncementHandler(AnnouncementHandler handler, String[] interfaces);

    /**
     * Start server mode.  The application creates the BusAttachment
     * @param port the bound AllJoyn session port. The application binds the port,
     *             and the about server announces it.
     * @param propertyStore a container of device/application properties.
     * @param m_bus the AllJoyn bus attachment.
     * @throws Exception indicating failure when starting the AboutServer
     * @see AboutKeys
     */
    public void startAboutServer(short port, PropertyStore propertyStore, BusAttachment m_bus) throws Exception;

    /**
     * Stop server mode.
     */
    public void stopAboutServer();

    /**
     * Add a BusObject and the BusInterfaces that it implements, to the server's
     * Announcement.
     * @param objPath the path of the BusObject
     * @param interfaces the BusInterfaces that the BusObject implements.
     * @see BusObjectDescription
     * @see AboutTransport#Announce(short, short, BusObjectDescription[], java.util.Map)
     */
    public void addObjectDescription(String objPath, String [] interfaces);

    /**
     * Remove the BusInterfaces that the given BusObject implements, from the
     * server's Announcement
     * @param objPath the path of the BusObject
     * @param interfaces the interfaces to remove.
     *                   Note: only the passed interfaces will be removed.
     */
    public void removeObjectDescription(String objPath, String [] interfaces);

    /**
     * A plural call of addBusObjectDescription
     * @param addBusObjectDescriptions list of BusObjectDescriptions to add to the next Announce signal
     * @see #addObjectDescription(String, String[])
     */
    public void addObjectDescriptions(List<BusObjectDescription> addBusObjectDescriptions);

    /**
     * A plural call of removeBusObjectDescription
     * @param removeBusObjectDescriptions list of BusObjectDescriptions to remove from next the Announce signal
     * @see #removeObjectDescription(String, String[])
     */
    public void removeObjectDescriptions(List<BusObjectDescription> removeBusObjectDescriptions);

    /**
     * Make the server send an Announcement
     */
    public void announce();

    /**
     * Register the application icon with the server
     *
     * @param mimetype the MimeType for the image
     * @param url  A URL that contains the location of the icon hosted in the cloud.
     * @param content an array of bytes that represent an icon
     * @throws Exception indicating failure to register the icon.
     */
    public void registerIcon(String mimetype, String url, byte[] content) throws Exception;

    /**
     * Unregister the application icon
     */
    public void unregisterIcon();
}
